Running setup.py build in Windows in Admin prompt results in Access denied errors

Hi, 

I tried using this tool in an elevated command prompt, but I got some strange "Access is denied" and "undefined reference to f" errors. See image: 

![image](https://user-images.githubusercontent.com/6684637/33035467-02d8f556-cde0-11e7-853f-45ec52e2efa9.png)

If I go to the location of my "root" file, I'm able to build using the go build command in the image above. 

I'm on Python 2.7.11, 32 bit on Windows 8 

I built Go from source, and I'm on git hash 8f70e1f8a91db6c8a1bcc292766af95e52e8512b

I have gcc installed in MINGW32

This is what my setup.py looks like: 
![image](https://user-images.githubusercontent.com/6684637/33036399-19b142e4-cde3-11e7-8539-22c9bf579c33.png)

There are a ton of other arguments I didn't include. I have a template for setup.py that I generate for each package I make. 

I'm wondering what the best approach to trying to debug this issue is. I've tried looking at the files its complaining about, but I can't find all of them. Any tips? 

Thanks!

I played around with this a bit more. It appears the main issue is now the "undefined reference to __imp__Py..." errors. I noticed that one of the things that was getting included was "C:\Python27\PC". This folder actually doesn't exist in my installation. I've noticed it doesn't exist in a fresh installation on another machine either, so this doesn't seem quite right. I think the missing references are meant to be in a file called libpython27.a in the C:\Python27\libs folder, but adding "-IC:\Python27\libs" to CGO_CFLAGS didn't seem to help. Not sure what to try next...

I tried following the instructions in this article since it looks very promising, but it didn't help either: 
http://techtonik.rainforce.org/2008/01/compiling-python-extension-with-gcc.html

I ended up with "c:/mingw/bin/../lib/gcc/mingw32/5.3.0/../../../../mingw32/bin/ld.exe: cannot find -lpython27"

I added the -L and -l flags to CGO_LDFLAGS. I'm not sure exactly what the gcc call will look like, and if I faithfully reproduced it.

I tried applying (iteratively) the changes necessary to get the tests to pass, my progress is [here](https://github.com/asottile/setuptools-golang/compare/windows?expand=1)

Just wanted to confirm a few things about your setup before continuing as I'm getting into territory I don't fully understand :)

1. mingw64 (gcc): 64 bit or 32bit?
2. go: 64bit or 32bit?
3. cpython: compiled with mingw? or from python.org installers
4. cpython: 64bit or 32bit?

For me I have these setups:

1. mingw: 64bit
2. go: 64bit
3. from python.org installers (msvc)
4. 64bit

I'm getting to the following:

### python2.7.14

I'm currently getting to the following build error:

```
    $ CGO_CFLAGS='-Ic:\python27\include -Ic:\temp\pytest-16\venv0\venv\PC' CGO_LDFLAGS='-Lc:\python27\libs -lpython27' G
OPATH='c:\temp\tmp8eqhfa' go build -buildmode=c-shared -o 'c:\temp\pip-eimzs6-build\build\lib.win-amd64-2.7\sum.pyd'
    # github.com/asottile/fake
    C:\Temp\go-build209371238\b001\_x003.o:sum.c:(.text+0x69): undefined reference to `__imp_Py_InitModule4'
    collect2.exe: error: ld returned 1 exit status
```

which seems to be fixable via [this](https://stackoverflow.com/a/9673051/812183)

Adding that define, I can successfully build, but something seems not right with the shared objects created:

```
        if returncode is not None:
            if proc.returncode != returncode:
                raise AssertionError(
                    '{!r} returned {} (expected {})\nout:\n{}\nerr:\n{}\n'.format(
>                       cmd, proc.returncode, returncode, out, err,
                    )
                )
E               AssertionError: (u'c:\\temp\\pytest-19\\venv0\\venv\\Scripts\\python', u'-c', u'import sum; print(sum.su
m(1, 2))') returned -1073741819 (expected 0)
E               out:
E
E               err:
E               None
```

### python3.6.3

For whatever reason (?) I don't need to jump through the `-DMS_WIN64` hoop -- but I end up at the same failure state as python2.7

--- a/README.md
+++ b/README.md
@@ -1,4 +1,5 @@
 [![Build Status](https://travis-ci.org/asottile/setuptools-golang.svg?branch=master)](https://travis-ci.org/asottile/setuptools-golang)
+[![Build status](https://ci.appveyor.com/api/projects/status/j4i2pc4o7pby3wdn/branch/master?svg=true)](https://ci.appveyor.com/project/asottile/setuptools-golang/branch/master)
 [![Coverage Status](https://img.shields.io/coveralls/asottile/setuptools-golang.svg?branch=master)](https://coveralls.io/r/asottile/setuptools-golang)
 
 setuptools-golang
@@ -12,6 +13,12 @@ This requires golang >= 1.5.  It is currently tested against 1.6 and 1.7.
 
 This requires python >= 2.7.  It is currently tested against 2.7, 3.5, 3.6,
 and pypy.
+
+## Platform Support
+
+- linux
+- macos
+- win32 (32 bit cpython, 32 bit go 1.10+)
 
 ## Usage
 

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,0 +1,19 @@
+environment:
+    matrix:
+    -   PYTHON: 'C:\Python27'
+    -   PYTHON: 'C:\Python36'
+
+install:
+-   'SET PATH=%PYTHON%;%PYTHON%\Scripts;C:\MinGW\bin;C:\go-src\bin;%PATH%'
+-   'SET GOROOT=C:\go-src'
+-   'SET GOROOT_BOOTSTRAP=C:\go19-x86'
+-   'git clone --branch=go1.10beta1 --depth=1 https://github.com/golang/go C:\go-src'
+-   'cd C:\go-src\src && make.bat'
+-   'pip install pytest'
+
+# Not a C# project
+build: false
+
+test_script: pytest tests
+
+cache: '%LOCALAPPDATA%\pip\cache'

--- a/setuptools_golang.py
+++ b/setuptools_golang.py
@@ -2,10 +2,12 @@ from __future__ import print_function
 from __future__ import unicode_literals
 
 import contextlib
+import copy
 import io
 import os
 import pipes
 import shutil
+import stat
 import subprocess
 import sys
 import tempfile
@@ -19,11 +21,16 @@ def _tmpdir():
     try:
         yield tempdir
     finally:
-        shutil.rmtree(tempdir)
+        def err(action, name, exc):  # pragma: no cover (windows)
+            """windows: can't remove readonly files, make them writeable!"""
+            os.chmod(name, stat.S_IWRITE)
+            action(name)
+
+        shutil.rmtree(tempdir, onerror=err)
 
 
 def _get_cflags(compiler):
-    return ' '.join('-I{}'.format(p) for p in compiler.include_dirs)
+    return str(' ').join(str('-I{}').format(p) for p in compiler.include_dirs)
 
 
 LFLAG_CLANG = '-Wl,-undefined,dynamic_lookup'
@@ -35,6 +42,12 @@ def _get_ldflags():
     """Determine the correct link flags.  This attempts dummy compiles similar
     to how autotools does feature detection.
     """
+    # windows gcc does not support linking with unresolved symbols
+    if sys.platform == 'win32':  # pragma: no cover (windows)
+        prefix = getattr(sys, 'real_prefix', sys.prefix)
+        libs = os.path.join(prefix, str('libs'))
+        return str('-L{} -lpython{}{}').format(libs, *sys.version_info[:2])
+
     cc = subprocess.check_output(('go', 'env', 'CC')).decode('UTF-8').strip()
 
     with _tmpdir() as tmpdir:
@@ -75,7 +88,13 @@ def _get_build_extension_method(base, root):
 
         # If there are no .go files then the parent should handle this
         if not any(source.endswith('.go') for source in ext.sources):
-            return base.build_extension(self, ext)
+            # the base class may mutate `self.compiler`
+            compiler = copy.deepcopy(self.compiler)
+            self.compiler, compiler = compiler, self.compiler
+            try:
+                return base.build_extension(self, ext)
+            finally:
+                self.compiler, compiler = compiler, self.compiler
 
         if len(ext.sources) != 1:
             _raise_error(
@@ -96,13 +115,13 @@ def _get_build_extension_method(base, root):
             shutil.copytree('.', root_path)
             pkg_path = os.path.join(root_path, main_dir)
 
-            env = {'GOPATH': tempdir}
+            env = {str('GOPATH'): tempdir}
             cmd_get = ('go', 'get', '-d')
             _check_call(cmd_get, cwd=pkg_path, env=env)
 
             env.update({
-                'CGO_CFLAGS': _get_cflags(self.compiler),
-                'CGO_LDFLAGS': _get_ldflags(),
+                str('CGO_CFLAGS'): _get_cflags(self.compiler),
+                str('CGO_LDFLAGS'): _get_ldflags(),
             })
             cmd_build = (
                 'go', 'build', '-buildmode=c-shared',

--- a/tests/setuptools_golang_test.py
+++ b/tests/setuptools_golang_test.py
@@ -26,8 +26,8 @@ def run(*cmd, **kwargs):
     returncode = kwargs.pop('returncode', 0)
     proc = subprocess.Popen(cmd, **kwargs)
     out, err = proc.communicate()
-    out = out.decode('UTF-8') if out is not None else None
-    err = err.decode('UTF-8') if err is not None else None
+    out = out.decode('UTF-8').replace('\r', '') if out is not None else None
+    err = err.decode('UTF-8').replace('\r', '') if err is not None else None
     if returncode is not None:
         if proc.returncode != returncode:
             raise AssertionError(
@@ -56,9 +56,10 @@ def venv(tmpdir_factory):
     """A shared virtualenv fixture, be careful not to install two of the same
     package into this -- or sadness...
     """
+    bin = 'Scripts' if sys.platform == 'win32' else 'bin'
     venv = tmpdir_factory.mktemp('venv').join('venv')
-    pip = venv.join('bin/pip').strpath
-    python = venv.join('bin/python').strpath
+    pip = venv.join(bin, 'pip').strpath
+    python = venv.join(bin, 'python').strpath
     # Make sure this virtualenv has the same executable
     run('virtualenv', venv.strpath, '-p', sys.executable)
     # Install this so we can get coverage
@@ -74,9 +75,9 @@ SUM = 'import {0}; print({0}.sum(1, 2))'
 @pytest.mark.parametrize(
     ('pkg', 'mod'),
     (
-        ('testing/sum', 'sum'),
-        ('testing/sum_pure_go', 'sum_pure_go'),
-        ('testing/sum_sub_package', 'sum_sub_package.sum'),
+        (os.path.join('testing', 'sum'), 'sum'),
+        (os.path.join('testing', 'sum_pure_go'), 'sum_pure_go'),
+        (os.path.join('testing', 'sum_sub_package'), 'sum_sub_package.sum'),
     ),
 )
 def test_sum_integration(venv, pkg, mod):
@@ -90,7 +91,8 @@ HELLO_WORLD = 'import project_with_c; print(project_with_c.hello_world())'
 
 def test_integration_project_with_c(venv):
     test_sum_integration(
-        venv, 'testing/project_with_c', 'project_with_c_sum.sum',
+        venv,
+        os.path.join('testing', 'project_with_c'), 'project_with_c_sum.sum',
     )
     out = run_output(venv.python, '-c', HELLO_WORLD)
     assert out == 'hello world\n'
@@ -100,14 +102,14 @@ RED = 'import red; print(red.red(u"ohai"))'
 
 
 def test_integration_imports_gh(venv):
-    run(venv.pip, 'install', 'testing/imports_gh')
+    run(venv.pip, 'install', os.path.join('testing', 'imports_gh'))
     out = run_output(venv.python, '-c', RED)
     assert out == '\x1b[31mohai\x1b[0m\n'
 
 
 def test_integration_notfound(venv):
     ret = run(
-        venv.pip, 'install', 'testing/notfound',
+        venv.pip, 'install', os.path.join('testing', 'notfound'),
         returncode=None, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
     )
     assert ret.returncode != 0
@@ -119,7 +121,7 @@ def test_integration_notfound(venv):
 
 def test_integration_multidir(venv):
     ret = run(
-        venv.pip, 'install', 'testing/multidir',
+        venv.pip, 'install', os.path.join('testing', 'multidir'),
         returncode=None, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
     )
     assert ret.returncode != 0
@@ -133,6 +135,6 @@ OHAI = 'import hello_lib; print(hello_lib.ohai(u"Anthony"))'
 
 
 def test_integration_internal_imports(venv):
-    run(venv.pip, 'install', 'testing/internal_imports')
+    run(venv.pip, 'install', os.path.join('testing', 'internal_imports'))
     out = run_output(venv.python, '-c', OHAI)
     assert out == 'ohai, Anthony\n'

